[IMP] reference: rewrite and move views documentation to user_interface (#3523)
odoo/documentation: content/developer/reference/user_interface/views/include/form.rst

.. _reference/user_interface/views/form:

Form
====

Form views are used to display the data from a single record. Their root element is
``<form>``. They are composed of regular HTML_ with additional structural and semantic
components.

.. code-block:: xml

  <form>
    ...
  </form>

Optional attributes_ are added on root element ``<form>`` to customize the view.

:string:
  string_

>   View name

Reviewer: used where and for what ?

:create:
  boolean_

  Disable/enable record creation on the view.

:edit:
  boolean_

  Disable/enable record editing on the view.

:duplicate:
  boolean_

  Disable/enable record duplication on the view through the **Action** dropdown.

:delete:
  boolean_

  Disable/enable record deletion on the view through the **Action** dropdown.

> :js_class:
>   string_
> 
>   Name of the javascript class to use when instantiating the view.
> 

Reviewer: ```suggestion
:js_class:
  string_

  Name of the javascript component the webclient will instantiating instead of the form the view.

```

> :disable_autofocus:
>   boolean_
> 
>   Disable automatic focus of the first field in the view.

Reviewer: ```suggestion
:disable_autofocus:
  boolean_

  Disable automatic focussing of the first field in the view.
```

.. _reference/user_interface/views/form/semantic:

Semantic components
-------------------

> Semantic components tie into and allow interaction with the Odoo system.
> Placeholders are denoted in all caps.

Reviewer: ```suggestion
Semantic components tie into the Odoo system and allow interaction with it.
remark: (with the remark style)
Placeholders are denoted in all caps.
```

.. _reference/user_interface/views/form/field:

<field>: render formatted values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: xml

  <form>
    <field name="FIELD_NAME"/>
  </form>

renders (and allow editing of, possibly) a single field of the current
record. Using several times a field in a form view is supported and the fields

> can receive different values for modifiers 'invisible' and 'readonly'. However,

Reviewer: shouldn't fields receive different values for 'invisible' so only one is visible at a time ?

for modifier 'required'. ``<field>`` can have the following attributes:

:name:
  string_ :ref:`model <reference/orm/model>` field name (mandatory)

  the name of the field to render

:id:
  string_

  the node id. Useful when there are several occurrences of the same field in
  the view (see ``label`` component below). Default is the field name.

:widget:
  string_

  fields have a default rendering based on their type
  (e.g. :class:`~odoo.fields.Char`, :class:`~odoo.fields.Many2one`).

  The ``widget`` attributes allows using a different rendering method and context.
  See more information in :ref:`reference/js/widgets`

  .. code-block:: xml

    <field name="tag_ids" widget="many2many_tags"/>

:options:
  :ref:`python expression <user_interface/views/python_expression>` that defines a dict_

  JSON object specifying configuration option for the field's widget
  (including default widgets)

  .. code-block:: xml

    <field name="tag_ids" widget="many2many_tags" options="{'color_field': 'FIELD_NAME', 'no_quick_create': True}"/>

:groups:
  `Comma-separated values`_ whose choices are the :class:`~odoo.addons.base.models.res_users.Groups` reference

>   only displays the field for specific users

Reviewer: could you add/link to the syntax of the groups: what does ! does, how to find the actual string that represent the group (AFAIR it's not the name but the xmlid) etc. pls :)

  .. code-block:: xml

    <field name="fname" groups="base.group_no_one,!base.group_multi_company"/>

:domain:
  :ref:`python expression <user_interface/views/python_expression>` that defines a :ref:`reference/orm/domains`

  for relational fields only, filters to apply when displaying existing
  records for selection

  .. code-block:: xml

    <field name="fname_id" domain="[('fname_a', '=', parent.fname_b)]"/>

:context:
  :ref:`python expression <user_interface/views/python_expression>` [#python_expression] that defines a dict_

  for relational fields only, context to pass when fetching possible values

  .. code-block:: xml

    <field name="fname" context="{
        'TYPE_view_ref': 'ADDON.MODEL_view_TYPE',
        'group_by': 'FIELD_NAME',
        'default_FIELD_NAME': ANY,
        'search_default_FIELD_NAME': True,
        'OTHER_BUSINESS_KEY': ANY,

>       }"/>

Reviewer: is there some place where we explain the *special* behavior of certain context keys, like `default_* ?
`

  :ref:`python expression <user_interface/views/python_expression>` that defines a boolean_

  standard dynamic attributes based on record values. If the value is
  trully or if the evaluate expression is trully, display the field in both
  readonly and edit mode, but never make it editable.

  .. code-block:: xml

    <field name="fname_a" readonly="True"/>
    <field name="fname_b" readonly="name_a in [fname_b, parent.fname_d]"/>

:required:
  :ref:`python expression <user_interface/views/python_expression>` that defines a boolean_

  standard dynamic attributes based on record values. If the value is
  trully or if the evaluate expression is trully, generates an error and
  prevents saving the record if the field doesn't have a value.

  .. code-block:: xml

    <field name="fname_a" required="True"/>
    <field name="fname_b" required="fname_c != 3 and fname_a == parent.fname_d"/>

:invisible:
  :ref:`python expression <user_interface/views/python_expression>` that defines a boolean_

  standard dynamic attributes based on record values. Hide the field
  if trully or if the evaluate expression is trully.

  .. code-block:: xml

    <field name="fname_a" invisible="True"/>

>     <field name="fname_b" invisible="fname_c != 3 and fname_a == parent.fname_d"/>

Reviewer: could you explain why it is useful to have invisible fields on a view, pls

:nolabel:
  boolean_

>   don't automatically display the field's label, only makes sense if the

Reviewer: ```suggestion
  if True, do not automatically display the field's label, only makes sense if the
```

:help:
  string_

  tooltip displayed for users when hovering the field or its label

:class:
  string_ `HTML class`_

  `HTML class`_ to set on the generated element.

  The styling use the Bootstrap_ framework and :ref:`UI icons <reference/user_interface/icons>`.

  Below are the common Odoo_ classes:

  ``oe_inline``: prevent the usual line break following fields and limit their span.

  ``oe_left``, ``oe_right``: floats_ the field to the corresponding direction

  ``oe_read_only``, ``oe_edit_only``: only displays the field in the corresponding form mode

  ``oe_avatar``: for image fields, displays images as "avatar" (square, 90x90 maximum size, some image decorations)

:filename:
  string_

  for binary fields, name of the related field providing the name of the file

:password:
  boolean_

  indicates that a :class:`~odoo.fields.Char` field stores a password and
  that its data shouldn't be displayed

:kanban_view_ref:
  string_ defined by the pattern: ``%(ADDON.MODEL_view_TYPE)s`` (target the :ref:`view reference <reference/view_record>`)

  for opening specific kanban view when selecting records from m2o/m2m in mobile
  environment

:default_focus:
  boolean_

>   defines the field on which the focus will be made when displaying the form view

Reviewer: ```suggestion
  If True, defines that this field is the fields that will be focussed when the view opens. Cannot be present on more than one field of a view.
```

:col:
  integer_ (default: 2)

  number of columns in a ``<group>``

:colspan:
  integer_ (default: 1)

  number of columns taken by an element

:string:
  string_

  displayed a group’s title

:invisible:
  same as for :ref:`field <reference/user_interface/views/form/field>` component.

Below is a possible structure and the representation of its rendering.

.. container:: row

  .. code-block:: xml
    :class: col-xxl-6

    <group>
      <field name="a" string="custom"/>
      <field name="b"/>
    </group>
    <group string="title 1">
      <group string="title 2">
        <field name="c"/>
        <field name="d"/>
      </group>
      <group>
        <field name="e"/>
        <field name="f"/>
        <field name="g"/>
      </group>
    </group>
    <group col="12">
      <group colspan="8">
        <field name="h"/>
      </group>
      <group colspan="4">
        <field name="i"/>
      </group>
    </group>

  .. image:: views/form_group.svg
    :class: col-xxl-6

.. _reference/user_interface/views/form/sheet:

<sheet>: Responsive layout
~~~~~~~~~~~~~~~~~~~~~~~~~~

``<sheet>`` can be used as a direct child to ``<form>`` for a narrower and more responsive

> form layout. Usually it contains :ref:`<group> <reference/user_interface/views/form/group>`.

Reviewer: can you explain what sheet does to make the form more responsive ?
